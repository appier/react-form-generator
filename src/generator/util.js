import {
  importComponent,
  embedSectionComponent,
  embedFieldComponent,
  importCSSModule,
  passProps,
  defaultProps,
} from './codeSnippet';

import prettier from 'prettier';

export const beautifier = template => {
  return prettier.format(template, {
    printWidth: 80,
    tabWidth: 2,
    singleQuote: true,
    trailingComma: true,
    bracketSpacing: true,
    jsxBracketSameLine: false,
    parser: 'babylon',
  });
};

export const camcelCaseToTemplateStr = s =>
  '__'
    .concat(s.replace(/(?:^|\.?)([A-Z])/g, (x, y) => '_' + y).toUpperCase())
    .concat('__');

export const replaceParam = params => text => {
  return Object.keys(params).reduce(
    (acc, key) => {
      const replacedStr = camcelCaseToTemplateStr(key);
      return acc.replace(new RegExp(`${replacedStr}`, 'g'), params[key]);
    },
    text,
  );
};

export const substitueParam = ({ componentName, params }) => template => {
  let ret = template.replace(/__COMPONENT_NAME__/g, componentName);
  ret = replaceParam(params)(ret);
  return ret;
};

export const getChildrenKeys = data => {
  return data.reduce(
    (acc, { type, key, children }) => {
      if (type === 'field') {
        acc.push(key);
        return acc;
      } else if (type === 'section') {
        return acc.concat(getChildrenKeys(children));
      }
    },
    [],
  );
};

export const getChildrenKeyPath = data => {
  return data.reduce(
    (acc, { type, key, keyPath, children }) => {
      if (type === 'field') {
        if (keyPath) {
          acc.push(keyPath);
        } else {
          acc.push([key]);
        }
        return acc;
      } else if (type === 'section') {
        return acc.concat(getChildrenKeyPath(children));
      }
    },
    [],
  );
};

export const getChildrenDefaultValue = data => {
  return data.reduce(
    (acc, { type, defaultValue, children }) => {
      if (type === 'field') {
        acc.push(defaultValue);
        return acc;
      } else if (type === 'section') {
        return acc.concat(getChildrenDefaultValue(children));
      }
    },
    [],
  );
};

export const injectChildrenComponent = children => template => {
  const importChildrenFileText = children
    .map(({ type, componentName }) => {
      if (type === 'field') {
        return importComponent({
          name: componentName,
          path: `./${componentName}`,
        });
      } else if (type === 'section') {
        return importComponent({
          name: componentName,
          path: `./${componentName}/Index`,
        });
      }
      return '';
    })
    .join('\n');

  const usedProps = getChildrenKeys(children).concat([
    'update',
    'registerValidator',
  ]);
  const usedPropsDefaultValue = getChildrenDefaultValue(children).concat([
    () => {},
    () => {},
  ]);

  const childrenComponentText = children
    .map(({ type, componentName, key, children }) => {
      if (type === 'field') {
        return embedFieldComponent({
          componentName,
          key,
          registerValidator: 'registerValidator',
          changeHandler: 'update',
        });
      } else if (type === 'section') {
        return embedSectionComponent({
          componentName,
          keys: getChildrenKeys(children),
          registerValidator: 'registerValidator',
          changeHandler: 'update',
        });
      }
      return '';
    })
    .join('\n');

  let ret = template
    .replace(/\/\* __IMPORT_COMPONENT__ \*\//g, importChildrenFileText)
    .replace(/{\/\* __INJECT_COMPONENT__ \*\/}/g, '\n' + childrenComponentText);

  ret = injectPassProps(usedProps)(ret);
  ret = setDefaultProps(usedProps, usedPropsDefaultValue)(ret);
  return ret;
};

export const injectPassProps = props => template => {
  let ret = template.replace(
    /\/\* __PASS_PROPS_TO_CHILDREN__ \*\//g,
    '\n' + passProps(props),
  );
  return ret;
};

export const setDefaultProps = (props, defaultValue) => template => {
  let ret = template.replace(
    /\/\* __PASS_PROPS__ \*\//g,
    '\n' + defaultProps(props, defaultValue),
  );
  return ret;
};

export const getSourceCSSFileLocation = rootPath => sourceCSS => {
  if (rootPath.length) {
    return [...rootPath, sourceCSS].join('/');
  } else {
    return './' + sourceCSS;
  }
};

export const injectCSSModule = config => rootPath => sourceCSS => template => {
  let ret = template;
  if (config.cssModule) {
    ret = ret.replace(
      /\/\* __IMPORT__ \*\//g,
      importCSSModule(getSourceCSSFileLocation(rootPath)(sourceCSS)),
    );
    ret = bindCSSModule(ret);
  } else {
    ret = ret.replace(/\s+\/\* __IMPORT__ \*\//g, '\n');
  }
  return ret;
};

export const bindCSSModule = template => {
  return template.replace(/className="([a-zA-Z\-_ ]+)"/g, (match, group) => {
    const cx = group.split(' ').map(d => `'${d}'`).join(', ');
    return `className={cx(${cx})}`;
  });
};