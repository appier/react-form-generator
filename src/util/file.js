import fs from 'fs-extra';

export const readFile = path => {
  return new Promise((resolve, reject) => {
    fs.readFile(path, 'utf8', (error, data) => {
      if (error)  reject(error)
      else        resolve(data)
    })
  })
}

export const writeFile = path => text => {
  return new Promise((resolve, reject) => {
    fs.outputFile(path, text, 'utf8', (error) => {
      if (error)  reject(error)
      else        resolve()
    });
  })
}

export const cleanDir = path => {
  return new Promise((resolve, reject) => {
    fs.emptyDir(path, (error) => {
      if (error)  reject(error)
      else        resolve()
    })
  })
}

export const copyFile = source => target => {
  return new Promise((resolve, reject) => {
    fs.copy(source, target, (error) => {
      if (error)  reject(error)
      else        resolve()
    })
  })
}


