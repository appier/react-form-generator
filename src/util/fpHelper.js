//monad
export const Box = x =>
({
  ap: other => other.map(x),
  chain: f => f(x),
  map: f => Box(f(x)),
  fold: f => f(x),
  inspect: () => `Box(${x})`
})

export const Right = x =>
({
  chain: f => f(x),
  ap: other => other.map(x),
  traverse: (of, f) => f(x).map(Right),
  map: f => Right(f(x)),
  fold: (f, g) => g(x),
  inspect: () => `Right(${x})`
})

export const Left = x =>
({
  chain: f => Left(x),
  ap: other => Left(x),
  traverse: (of, f) => of(Left(x)),
  map: f => Left(x),
  fold: (f, g) => f(x),
  inspect: () => `Left(${x})`
})

export const fromNullable = x =>
  x != null ? Right(x) : Left(null)

export const tryCatch = f => {
  try {
    return Right(f())
  } catch(e) {
    return Left(e)
  }
}


// helper
export const True = () => true;
export const False = () => false;

export const compose = (...fns) => x => fns.reduceRight((v, f) => f(v), x);

export const curry = (f, ...args) =>
  (f.length <= args.length) ? f(...args) : (...more) => curry(f, ...args, ...more);

export const flip = fn => (...fns) => fn(...fns.reverse());

export const pipe = flip(compose);

export const ifElse = pred => onTrue => onFalse => d => pred(d) ? onTrue(d) : onFalse(d);
export const identity = d => d;
export const thunk = d => () => d;
export const first = d => d[0];
export const equal = a => b => a === b;
export const not = fn => x => !fn(x);
export const and = fa => fb => x => fa(x) && fb(x);
export const or = fa => fb => x => fa(x) || fb(x);
export const xor = fa => fb => x =>  (fa(x) && !fb(x)) || (!fa(x) && fb(x));
export const props = field => x => x[field];
export const map = fn => x => x.map(fn);
export const filter = fn => x => x.filter(fn);
export const reduce = fn => init => x => x.reduce((acc, d)=>fn(acc)(d), init);
export const some = fn => x => x.some(fn);
export const every = fn => x => x.every(fn);


export const peek = tag => d => {
  const msg = typeof d === 'object' ? JSON.stringify(d, null, 2) : String(d);
  console.log(tag, msg);
  return d;
}
