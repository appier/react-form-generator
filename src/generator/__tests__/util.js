import { bindCSSModule, camcelCaseToTemplateStr } from '../util';

describe('codeSnippet Generator', () => {
  it('bindCSSModule', () => {
    expect(
      bindCSSModule('<div className="vertical-input-group __CLASS_NAME__">'),
    ).toMatchSnapshot();
    expect(bindCSSModule('<div className="input-wrap">')).toMatchSnapshot();
    expect(bindCSSModule('className="input"')).toMatchSnapshot();
  });

  it('camcelCaseToTemplateStr', () => {
    expect(camcelCaseToTemplateStr('className')).toMatchSnapshot();
    expect(camcelCaseToTemplateStr('label')).toMatchSnapshot();
    expect(camcelCaseToTemplateStr('placeholder')).toMatchSnapshot();
    expect(camcelCaseToTemplateStr('componentName')).toMatchSnapshot();
  });
  // it('indentText', () => {
  //   expect(indentText(0)('hello')).toMatchSnapshot();
  //   expect(indentText(1)('hello')).toMatchSnapshot();
  //   expect(indentText(1)('  hello')).toMatchSnapshot();
  //   expect(indentText(4)('  hello')).toMatchSnapshot();
  // });
});