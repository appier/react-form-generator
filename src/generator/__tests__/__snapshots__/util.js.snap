exports[`codeSnippet Generator bindCSSModule 1`] = `"<div className={cx(\'vertical-input-group __CLASS_NAME__\')}>"`;

exports[`codeSnippet Generator bindCSSModule 2`] = `"<div className={cx(\'input-wrap\')}>"`;

exports[`codeSnippet Generator bindCSSModule 3`] = `"className={cx(\'input\')}"`;

exports[`codeSnippet Generator camcelCaseToTemplateStr 1`] = `"__CLASS_NAME__"`;

exports[`codeSnippet Generator camcelCaseToTemplateStr 2`] = `"__LABEL__"`;

exports[`codeSnippet Generator camcelCaseToTemplateStr 3`] = `"__PLACEHOLDER__"`;

exports[`codeSnippet Generator camcelCaseToTemplateStr 4`] = `"__COMPONENT_NAME__"`;
